### ISSUE TITLE
False negative: potentially unbound instance attribute doesn't trigger `possibly-undefined` error

### ISSUE THREAD

**Issue by nsoranzo** [python/mypy#17837]
**Bug Report**

When enabling the [possibly-undefined](https://mypy.readthedocs.io/en/stable/error_code_list2.html#warn-about-variables-that-are-defined-only-in-some-execution-paths-possibly-undefined) error code, a potentially undefined instance attribute doesn't trigger the error, while a potentially unbound variable does.

**To Reproduce**

```python
def test_possibly_undefined(flag: bool):
    if flag:
        a = 1
    print(a)  # error: Name "a" may be undefined  [possibly-undefined]


class Foo:
    def test_possibly_undefined(self, flag: bool):
        if flag:
            self.a = 1
        print(self.a)  # false negative, should also be a possibly-undefined error


try:
    test_possibly_undefined(False)
except Exception as e:
    print(f"{type(e).__name__}: {e}")  # UnboundLocalError: cannot access local variable 'a' where it is not associated with a value

f = Foo()
try:
    f.test_possibly_undefined(False)
except Exception as e:
    print(f"{type(e).__name__}: {e}")  # AttributeError: 'Foo' object has no attribute 'a'
```

Also available as [mypy Playground](https://mypy-play.net/?mypy=latest&python=3.12&gist=a4719f9bab7c797f4fa9cec8b356515e), but cannot enable `enable_error_code = possibly-undefined` there.

**Expected Behavior**

A `possible-undefined` error should be reported for line 11 as well, e.g.:

```
test.py:4: error: Name "a" may be undefined  [possibly-undefined]
test.py:11: error: Attribute "a" may be undefined  [possibly-undefined]
Found 2 errors in 1 file (checked 1 source file)
```

**Actual Behavior**

No error reported for line 11.

```
test.py:4: error: Name "a" may be undefined  [possibly-undefined]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.11.2
- Mypy command-line flags: `--enable-error-code=possibly-undefined`
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.12.6

**Comment by brianschubert** [comment]
Related: #4019, #10736

Tracking conditionally defined attributes is hard, both for mypy and for humans. Generally it's preferable to have attributes always be defined, and if necessary give them a default value / optional type that can be guarded against. Though that wouldn't have helped in the linked issue, since the attribute wasn't supposed to be conditionally defined to begin with.

This sort of error can be easier to catch against if you used a pattern like
```
if x:
    foo = 1
elif y:
    foo = 2

self.foo = foo
```
or
```
self.foo = _resolve_foo(x, y)
```

**Comment by nsoranzo** [comment]
@brianschubert Thanks for the quick reply and the links. Agreed that it's hard, that's why we need a tool like mypy to help :)

As you can guess, in a project with hundreds of contributors it is nearly impossible to ensure the use of such patterns every time an instance attribute is initialised.
It's OK if there's no plan to address this, I think it's worth to have an issue specifically related to the the `possibly-undefined` error code.
